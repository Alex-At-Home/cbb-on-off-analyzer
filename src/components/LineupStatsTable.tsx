// React imports:
import React, { useState, useEffect } from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';

// Additional components:
// @ts-ignore
import LoadingOverlay from 'react-loading-overlay';
import Select, { components} from "react-select";

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "./GenericTable";
import { LineupFilterParams, ParamDefaults } from '../utils/FilterModels';
import { RosterStatsModel } from '../components/RosterStatsTable';

// Util imports
import { CbbColors } from "../utils/CbbColors";
import { CommonTableDefs } from "../utils/CommonTableDefs";
import { PositionUtils } from "../utils/stats/PositionUtils";

export type LineupStatsModel = {
  lineups?: Array<any>,
  error_code?: string
}
type Props = {
  lineupStats: LineupStatsModel,
  rosterStats: RosterStatsModel,
  startingState: LineupFilterParams;
  onChangeState: (newParams: LineupFilterParams) => void;
}

const LineupStatsTable: React.FunctionComponent<Props> = ({lineupStats, rosterStats, startingState, onChangeState}) => {

  // 1] State

  const [ minPoss, setMinPoss ] = useState(startingState.minPoss || ParamDefaults.defaultLineupMinPos);
  const [ maxTableSize, setMaxTableSize ] = useState(startingState.maxTableSize || ParamDefaults.defaultLineupMaxTableSize);
  const [ sortBy, setSortBy ] = useState(startingState.sortBy || ParamDefaults.defaultLineupSortBy);
  const [ filterStr, setFilterStr ] = useState(startingState.filter || ParamDefaults.defaultLineupFilter);

  // (slight delay when typing into the filter to make it more responsive)
  const [ timeoutId, setTimeoutId ] = useState(-1);
  const [ tmpFilterStr, setTmpFilterStr ] = useState(filterStr);

  const filterFragments =
    filterStr.split(",").map(fragment => _.trim(fragment)).filter(fragment => fragment ? true : false);
  const filterFragmentsPve =
    filterFragments.filter(fragment => fragment[0] != '-');
  const filterFragmentsNve =
    filterFragments.filter(fragment => fragment[0] == '-').map(fragment => fragment.substring(1));

  useEffect(() => { //(this ensures that the filter component is up to date with the union of these fields)
    const newState = _.merge(startingState, {
      minPoss: minPoss,
      maxTableSize: maxTableSize,
      sortBy: sortBy,
      filter: filterStr
    });
    onChangeState(newState);
  }, [ minPoss, maxTableSize, sortBy, filterStr ]);

  // 2] Data Model

  // 3] Utils

  // 3.1] Build individual info

  // 3.1.1] Positional info from the season stats

  const positionFromPlayerKey = _.chain(rosterStats.global || []).map((player: any) => {
    const [ posConfs, posConfsDiags ] = PositionUtils.buildPositionConfidences(player);
    const [ pos, posDiags ] = PositionUtils.buildPosition(posConfs, player);
    return [ player.key, { posConfidences: _.values(posConfs || {}), posClass: pos } ];
  }).fromPairs().value();

  // 3.2] Table building

  const offPrefixFn = (key: string) => "off_" + key;
  const offCellMetaFn = (key: string, val: any) => "off";
  const defPrefixFn = (key: string) => "def_" + key;
  const defCellMetaFn = (key: string, val: any) => "def";

  const sorter = (sortStr: string) => { // format: (asc|desc):(off_|def_|diff_)<field>
    const sortComps = sortStr.split(":"); //asc/desc
    const dir = (sortComps[0] == "desc") ? -1 : 1;
    const fieldComps = _.split(sortComps[1], "_", 1); //off/def/diff
    const fieldName = sortComps[1].substring(fieldComps[0].length + 1); //+1 for _
    const field = (lineup: any) => {
      switch(fieldComps[0]) {
        case "diff": //(off-def)
          return (lineup["off_" + fieldName]?.value || 0.0)
                - (lineup["def_" + fieldName]?.value || 0.0);
        default: return lineup[sortComps[1]]?.value; //(off or def)
      }
    };
    return (lineup: any) => {
      return dir*(field(lineup) || 0);
    };
  };

  const lineups = lineupStats?.lineups || [];
  const tableData = _.chain(lineups).filter((lineup) => {
      const minPossInt = parseInt(minPoss);
      const offPos = lineup.off_poss?.value || 0;
      const defPos = lineup.def_poss?.value || 0;
      return offPos >= minPossInt || defPos >= minPossInt; //(unclear which of || vs && is best...)
    }).filter((lineup) => {
      const namesToTest = _.chain(lineup?.players_array?.hits?.hits).flatMap((pDoc) => {
        return pDoc?._source.players.map((p: any) => p.id);
      }).value().concat([ lineup.key ]);

      const playerFilter =
        (_.isEmpty(filterFragmentsPve) ||
          _.every(filterFragmentsPve, (frag) => _.some(namesToTest, (name) => name.indexOf(frag) >= 0))
        ) &&
        (_.isEmpty(filterFragmentsNve) ||
          !_.some(filterFragmentsNve, (frag) => _.some(namesToTest, (name) => name.indexOf(frag) >= 0))
        );
      return playerFilter && (lineup.key != ""); // (workaround for #53 pending fix)

    }).sortBy(
       [ sorter(sortBy) ]
    ).take(
      parseInt(maxTableSize)
    ).flatMap((lineup) => {
      const codesAndIds = lineup.players_array?.hits?.hits?.[0]?._source?.players || [];

      const sortedCodesAndIds = PositionUtils.orderLineup(codesAndIds, positionFromPlayerKey);

      const title = sortedCodesAndIds.map((cid: { code: string, id: string}) => cid.code).join(" / ");
      const stats = { off_title: title, def_title: "", ...lineup };
      return [
        GenericTableOps.buildDataRow(stats, offPrefixFn, offCellMetaFn),
        GenericTableOps.buildDataRow(stats, defPrefixFn, defCellMetaFn),
        GenericTableOps.buildRowSeparator()
      ];
    }).value();

  // 3.2] Sorting utils

  const sortOptions: Array<any> = _.flatten(
    _.toPairs(CommonTableDefs.lineupTable)
      .filter(keycol => keycol[1].colName && keycol[1].colName != "")
      .map(keycol => {
        return [
          ["desc","off"], ["asc","off"], ["desc","def"], ["asc","def"], ["desc","diff"], ["asc","diff"]
        ].map(combo => {
          const ascOrDesc = (s: string) => { switch(s) {
            case "asc": return "Asc.";
            case "desc": return "Desc.";
          }}
          const offOrDef = (s: string) => { switch(s) {
            case "off": return "Offensive";
            case "def": return "Defensive";
            case "diff": return "Off-Def";
          }}
          return {
            label: `${keycol[1].colName} (${ascOrDesc(combo[0])} / ${offOrDef(combo[1])})`,
            value: `${combo[0]}:${combo[1]}_${keycol[0]}`
          };
        });
      })
  );
  const sortOptionsByValue = _.fromPairs(
    sortOptions.map(opt => [opt.value, opt])
  );
  /** Put these options at the front */
  const mostUsefulSubset = [
    "desc:off_poss",
    "desc:diff_adj_ppp",
    "desc:off_adj_ppp",
    "asc:def_adj_ppp",
  ];
  /** The two sub-headers for the dropdown */
  const groupedOptions = [
    {
      label: "Most useful",
      options: _.chain(sortOptionsByValue).pick(mostUsefulSubset).values().value()
    },
    {
      label: "Other",
      options: _.chain(sortOptionsByValue).omit(mostUsefulSubset).values().value()
    }
  ];
  /** The sub-header builder */
  const formatGroupLabel = (data: any) => (
    <div>
      <span>{data.label}</span>
    </div>
  );

  // 3] Utils
  function picker(offScale: (val: number) => string, defScale: (val: number) => string) {
    return (val: any, valMeta: string) => {
      const num = val.value as number;
      return _.isNil(num) ?
        CbbColors.malformedDataColor : //(we'll use this color to indicate malformed data)
        ("off" == valMeta) ? offScale(num) : defScale(num)
        ;
    };
  }
  /** Sticks an overlay on top of the table if no query has ever been loaded */
  function needToLoadQuery() {
    return lineupStats.lineups === undefined;
  }
  function rowSpanCalculator(cellMeta: string) {
    switch(cellMeta) {
      case "off": return 2;
      case "def": return 0;
      default: return 1;
    }
  }

  /** For use in selects */
  function stringToOption(s: string) {
    return sortOptionsByValue[s];
  }

  /** Handling filter change (/key presses to fix the select/delete on page load) */
  const onFilterChange = (ev: any) => {
    const toSet = ev.target.value;
    setTmpFilterStr(toSet);
    if (timeoutId != -1) {
      window.clearTimeout(timeoutId);
    }
    setTimeoutId(window.setTimeout(() => {
      setFilterStr(toSet);
    }, 100));
  };

  // 4] View
  return <Container>
    <LoadingOverlay
      active={needToLoadQuery()}
      text={lineupStats.error_code ?
        `Query Error: ${lineupStats.error_code}` :
        "Press 'Submit' to view results"
      }
    >
      <Form.Row>
        <Form.Group as={Col} sm="6">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="filter">Filter</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control
              onKeyUp={onFilterChange}
              onChange={onFilterChange}
              placeholder = "eg Player1Surname,Player2FirstName,-Player3Name"
              value={tmpFilterStr}
            />
          </InputGroup>
        </Form.Group>
      </Form.Row>
      <Form.Row>
        <Form.Group as={Col} sm="3">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="maxLineups">Max Lineups</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control
              onChange={(ev: any) => {
                if (ev.target.value.match("^[0-9]*$") != null) {
                  setMaxTableSize(ev.target.value);
                }
              }}
              placeholder = "eg 50"
              value={maxTableSize}
            />
          </InputGroup>
        </Form.Group>
        <Form.Group as={Col} sm="3">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="minPossessions">Min Poss #</InputGroup.Text>
            </InputGroup.Prepend>
            <Form.Control
              onChange={(ev: any) => {
                if (ev.target.value.match("^[0-9]*$") != null) {
                  setMinPoss(ev.target.value);
                }
              }}
              placeholder = "eg 5"
              value={minPoss}
            />
          </InputGroup>
        </Form.Group>
        <Form.Group as={Col} sm="6">
          <InputGroup>
            <InputGroup.Prepend>
              <InputGroup.Text id="sortBy">Sort By</InputGroup.Text>
            </InputGroup.Prepend>
            <Select
              className="w-75"
              value={ stringToOption(sortBy) }
              options={ groupedOptions }
              onChange={(option) => { if ((option as any)?.value)
                setSortBy((option as any)?.value);
              }}
              formatGroupLabel={formatGroupLabel}
            />
          </InputGroup>
        </Form.Group>
      </Form.Row>
      <Row>
        <Col>
          <GenericTable tableCopyId="lineupStatsTable" tableFields={CommonTableDefs.lineupTable} tableData={tableData}/>
        </Col>
      </Row>
    </LoadingOverlay>
  </Container>;
};

export default LineupStatsTable;
