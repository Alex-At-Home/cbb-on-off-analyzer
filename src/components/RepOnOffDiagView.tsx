// React imports:
import React, { useState } from 'react';

// Lodash
import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Library Utils
// @ts-ignore
import { mean, std } from "mathjs";

// Utils
import { OnOffReportDiagUtils } from "../utils/OnOffReportDiagUtils";

type Props = {
  diagInfo: Array<any>,
  player: Record<string, any>,
  expandedMode: boolean,
  showHelp: boolean
};

const RepOnOffDiagView: React.FunctionComponent<Props> = ({diagInfo, player, expandedMode, showHelp}) => {

  const doSame4Analysis = (offOrDef: "off" | "def", dir: 1 | -1) => {
    const field = "adjEff";
    const contrib = (lineupPlusDiag: any, field: string) =>
      (lineupPlusDiag?.contrib?.[offOrDef]?.[field] || 0) as number;
    const lineups = diagInfo.filter(
      (lineupPlusDiag) => contrib(lineupPlusDiag, field)*dir < 0
    );
    const contribs = lineups.map((lineupPlusDiag) => contrib(lineupPlusDiag, field));
    const sumContrib = _.sum(contribs)
    const meanContrib: any = contribs.length ? mean(contribs) : 0;
    const stdevContrib: any = contribs.length ? std(contribs) : 0;
    return [
      _.chain(lineups).filter( // has to be > 1x stdev
        (lineupPlusDiag) => Math.abs(contrib(lineupPlusDiag, field) - meanContrib) > 1*stdevContrib
      ).filter( // only care about 1 direction
        (lineupPlusDiag) => (dir == -1) == (contrib(lineupPlusDiag, field) > meanContrib)
      ).sortBy(
        [ (lineupPlusDiag) => dir*contrib(lineupPlusDiag, field) ]
      ).take(5).map((lineupPlusDiag, i) => {
        return <li key={"" + i}>
          <a href="#">{lineupPlusDiag.keyArray.join(" / ")}</a>:
          &nbsp;contrib=[<b>{contrib(lineupPlusDiag, field).toFixed(2)}</b>]
          &nbsp;(poss=[<b>{contrib(lineupPlusDiag, "totalPoss").toFixed(1)}</b>], [<b>{(100*contrib(lineupPlusDiag, "possWeight")).toFixed(1)}%</b>])
        </li>
      }).value(),
      sumContrib, meanContrib, stdevContrib, contribs.length
    ];
  };
  const peerInfo = OnOffReportDiagUtils.buildPeerStatsForPlayer(player, diagInfo);
  const doPeerAnalysis = (offOrDef: "off" | "def") => {
    return _.keys(peerInfo).join(" , "); //TODO
  };


  const buildInfo = (offOrDef: "off" | "def") => {

    const infoHtml = _.flatMap([ -1, 1 ], (dir: 1 | -1, index) => {
      const [
        keyLineupHtml, keySum, keyMean, keyStdev, keyLength
      ] = doSame4Analysis(offOrDef, dir);

      const goodOrBad = (dir > 0) == (offOrDef == "off") ? "Bad" : "Good";

      return [
        <li key={offOrDef + index}>[<b>{keyLength}</b>] {goodOrBad} "Same-4"s, total contrib=[<b>{keySum.toFixed(1)}</b>] (mean=[<b>{keyMean.toFixed(1)}</b>], std=[<b>{keyStdev.toFixed(1)}</b>]). Key lineups:</li>,
        <ul key={"ul" + offOrDef + index}>
          {keyLineupHtml}
        </ul>,
        <li key={"peer" + offOrDef + index}>TODO {doPeerAnalysis(offOrDef)}</li>
      ]
    });

    return <ul>
      {infoHtml}
    </ul>;
  }

  return <span>
    {OnOffReportDiagUtils.getTitle(player, showHelp)}
    <ul>
      <li>Offensive analysis:</li>
      {buildInfo("off")}
      <li>Defensive analysis:</li>
      {buildInfo("def")}
    </ul>
    {expandedMode ?
      <span>For a more detailed analysis click here to view just this player.</span>
      : null
    }
  </span>;
}

export default RepOnOffDiagView;
