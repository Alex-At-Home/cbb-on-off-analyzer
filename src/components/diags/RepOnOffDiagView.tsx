// React imports:
import React, { useState } from 'react';

// Lodash
import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Library Utils
// @ts-ignore
import { mean, std } from "mathjs";

// Utils
import { OnOffReportDiagUtils } from "../../utils/tables/OnOffReportDiagUtils";
import { CommonFilterParams } from "../../utils/FilterModels";
import { QueryUtils } from "../../utils/QueryUtils";

type Props = {
  diagInfo: Array<any>,
  player: Record<string, any>,
  playerMap: Record<string, string>, //(code -> id)
  commonParams: CommonFilterParams,
  expandedMode: boolean,
  onExpand: (playerId: string) => void,
  showHelp: boolean,
  keyLineupThreshold?: number //(for testing, leave blank in prod)
};

const RepOnOffDiagView: React.FunctionComponent<Props> = ({diagInfo, player, playerMap, commonParams, expandedMode, onExpand, showHelp, keyLineupThreshold}) => {

  const threshold = _.isNil(keyLineupThreshold) ? 1 : keyLineupThreshold;

  // Build player comparison links with peers:
  const baseMaybeAdvQuery = QueryUtils.extractAdvancedQuery(commonParams.baseQuery || "");
  const compareLinkFromPeer = (peerCode: string) => {
    return OnOffReportDiagUtils.buildPlayerComparisonLink(
      player.playerId, player.playerCode, playerMap[peerCode] || peerCode, peerCode, baseMaybeAdvQuery, commonParams
    );
  };

  const doSame4Analysis = (offOrDef: "off" | "def", dir: 1 | -1) => {
    const field = "adjEff";
    const contrib = (lineupPlusDiag: any, field: string) =>
      (lineupPlusDiag?.contrib?.[offOrDef]?.[field] || 0) as number;
    const lineups = diagInfo.filter(
      (lineupPlusDiag) => contrib(lineupPlusDiag, field)*dir < 0
    );
    const contribs = lineups.map((lineupPlusDiag) => contrib(lineupPlusDiag, field));
    const sumContrib = _.sum(contribs)
    const meanContrib: any = contribs.length ? mean(contribs) : 0;
    const stdevContrib: any = contribs.length ? std(contribs) : 0;
    const keyLineups =  _.chain(lineups).filter( // has to be > 1x stdev (unless threshold disabled)
        (lineupPlusDiag) => !threshold || (Math.abs(contrib(lineupPlusDiag, field) - meanContrib) > threshold*stdevContrib)
      ).filter( // only care about 1 direction
        (lineupPlusDiag) => (dir == -1) == (contrib(lineupPlusDiag, field) >= meanContrib)
      ).sortBy(
        [ (lineupPlusDiag) => dir*contrib(lineupPlusDiag, field) ]
      ).take(5).value();

    const keyPeers = _.chain(keyLineups).flatMap((v) => _.toPairs(v.peers)).transform((acc, kv) => {
        const peerId = kv[0] as string;
        const peerInfo = kv[1] as Record<string, any>;
        if (!acc.hasOwnProperty(peerId)) {
          acc[peerId] = peerInfo.poss as number;
        } else {
          acc[peerId] += peerInfo.poss as number;
        }
      }, {} as Record<string, number>).toPairs().sortBy([(kv) => -kv[1]]).take(5).value();

    return [
      keyLineups.map((lineupPlusDiag, i) => {
        const title = lineupPlusDiag.keyArray.join(" / ");
        const same4Ids = lineupPlusDiag.keyArray.map((code: string) => playerMap[code] || code);
        return <li key={"" + i}>
          {OnOffReportDiagUtils.buildOnOffAnalysisLink(player.playerId, same4Ids, commonParams, title)}
          &nbsp;contrib=[<b>{contrib(lineupPlusDiag, field).toFixed(2)}</b>]
          &nbsp;(poss=[<b>{contrib(lineupPlusDiag, "totalPoss").toFixed(1)}</b>], [<b>{(100*contrib(lineupPlusDiag, "possWeight")).toFixed(1)}%</b>])
        </li>
      }),
      keyPeers.map((peerKv, i) => {
        const append = (keyPeers.length - 1 == i) ? "" : ", ";
        return <span key={"peer" + i}>{compareLinkFromPeer(peerKv[0])} (poss=[<b>{peerKv[1].toFixed(0)}</b>]){append}</span>
      }),
      sumContrib, meanContrib, stdevContrib, contribs.length
    ];
  };

  const buildInfo = (offOrDef: "off" | "def") => {

    const infoHtml = _.flatMap([ -1, 1 ], (dir: 1 | -1, index) => {
      const [
        keyLineupHtml, keyPeerHtml, keySum, keyMean, keyStdev, keyLength
      ] = doSame4Analysis(offOrDef, dir);

      const goodOrBad = (dir > 0) == (offOrDef == "off") ? "Bad" : "Good";

      return [
        <li key={offOrDef + index}>[<b>{keyLength}</b>] {goodOrBad} "Same-4"s, total contrib=[<b>{keySum.toFixed(1)}</b>] (mean=[<b>{keyMean.toFixed(1)}</b>], std=[<b>{keyStdev.toFixed(1)}</b>]).
        {keyLineupHtml.length > 0 ? " Key lineups:" : " All lineups similar."}</li>,
        <ul key={"ul" + offOrDef + index}>
          {keyLineupHtml.length > 0 ? keyLineupHtml.concat(
            [<li key={"ul" + offOrDef + "peers"}>Key sub-ins: {keyPeerHtml}</li>]
          ) : null}
        </ul>
      ]
    });

    return <ul>
      {infoHtml}
    </ul>;
  }

  return <span>
    {OnOffReportDiagUtils.getTitle(player, showHelp)}
    <ul>
      <li>Offensive analysis ([<b>{player?.replacement?.off_adj_ppp?.value.toFixed(1)}</b>] pts/100):</li>
      {buildInfo("off")}
      <li>Defensive analysis ([<b>{player?.replacement?.def_adj_ppp?.value.toFixed(1)}</b>] pts/100):</li>
      {buildInfo("def")}
      {!expandedMode ?
        <li><a href="#" onClick={(event) => { event.preventDefault(); onExpand(player.playerCode) }}>To view more detailed analysis for just this player click here</a>.</li>
        :
        <li><a href="#" onClick={(event) => { event.preventDefault(); onExpand("") }}>To clear the detailed analysis and player filter click here</a>.</li>
      }
    </ul>
  </span>;
}

export default RepOnOffDiagView;
