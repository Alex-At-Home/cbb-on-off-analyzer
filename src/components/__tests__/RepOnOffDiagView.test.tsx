import renderer from 'react-test-renderer';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { OnOffReportDiagUtils } from '../../utils/stats/OnOffReportDiagUtils';
import { CommonFilterParams } from "../../utils/FilterModels";
import { LineupUtils } from "../../utils/stats/LineupUtils";
import { LineupStatsModel } from '../LineupStatsTable';
import { sampleLineupStatsResponse } from "../../sample-data/sampleLineupStatsResponse"

import RepOnOffDiagView from "../RepOnOffDiagView";

describe("RepOnOffDiagView", () => {

  const lineupReport = {
    lineups: sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets,
    avgEff: 100.0,
    error_code: "test"
  };
  const tempTeamReport = LineupUtils.lineupToTeamReport(
    lineupReport, true, 100, 10
  );
  const playersWithAdjEff = tempTeamReport.players || [];
  const diagInfo = OnOffReportDiagUtils.getRepOnOffDiagInfo(playersWithAdjEff[2], 100);

  [ true, false ].forEach((expandedMode) => {
    const threshold = expandedMode ? 0 : undefined; //test both with default threshold and with 0
    test(`RepOnOffDiagView - should create snapshot (expanded mode=[${expandedMode}], threshold=[${threshold}])`, () => {
      const component = renderer.create(<RepOnOffDiagView
        diagInfo={diagInfo || []}
        player={playersWithAdjEff?.[0]}
        playerMap={{"AaWiggins": "Wiggins, Aaron"}}
        expandedMode={expandedMode}
        commonParams={{baseQuery: "test"}}
        onExpand={(playerId: string) => false}
        showHelp={true}
        keyLineupThreshold={threshold}
      />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
