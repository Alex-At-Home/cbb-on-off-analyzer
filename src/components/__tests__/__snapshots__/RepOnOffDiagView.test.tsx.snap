// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepOnOffDiagView RepOnOffDiagView - should create snapshot (expanded mode=[false], threshold=[undefined]) 1`] = `
<span>
  <span>
    <b>
      Replacement 'On-Off' 
      
      Diagnostics For [
      Wiggins, Aaron
      ]
    </b>
     
    <a
      href="https://hoop-explorer.blogspot.com/2020/03/diagnostics-mode-for-replacement-onoff.html"
      target="_new"
    >
      (?)
    </a>
  </span>
  <ul>
    <li>
      Offensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          0.8
        </b>
        ] (mean=[
        <b>
          0.8
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          -5.6
        </b>
        ] (mean=[
        <b>
          -5.6
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      Defensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          3.8
        </b>
        ] (mean=[
        <b>
          3.8
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
      <li>
        [
        <b>
          0
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          0.0
        </b>
        ] (mean=[
        <b>
          0.0
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      <a
        href="#"
        onClick={[Function]}
      >
        To view more detailed analysis for just this player click here
      </a>
      .
    </li>
  </ul>
</span>
`;

exports[`RepOnOffDiagView RepOnOffDiagView - should create snapshot (expanded mode=[true], threshold=[0]) 1`] = `
<span>
  <span>
    <b>
      Replacement 'On-Off' 
      
      Diagnostics For [
      Wiggins, Aaron
      ]
    </b>
     
    <a
      href="https://hoop-explorer.blogspot.com/2020/03/diagnostics-mode-for-replacement-onoff.html"
      target="_new"
    >
      (?)
    </a>
  </span>
  <ul>
    <li>
      Offensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          0.8
        </b>
        ] (mean=[
        <b>
          0.8
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         Key lineups:
      </li>
      <ul>
        <li>
          <div>
            <a
              href="/?autoOffQuery=false&baseQuery=test&offQuery=NOT%20%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&onQuery=%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&"
              target="_blank"
            >
              AaWiggins / AnCowan / ErAyala / JaSmith
            </a>
          </div>
           contrib=[
          <b>
            0.79
          </b>
          ]  (poss=[
          <b>
            134.2
          </b>
          ], [
          <b>
            39.4
            %
          </b>
          ])
        </li>
        <li>
          Key sub-ins: 
          <span>
            <div>
              <a
                href="/TeamReport?baseQuery=%28%7B%22Wiggins%2C%20Aaron%22%3B%22DoScott%22%7D%3D1%29%20AND%20%28test%29&filter=AaWiggins&incRepOnOff=true&repOnOffDiagMode=20&"
                target="_blank"
              >
                DoScott
              </a>
            </div>
             (poss=[
            <b>
              102
            </b>
            ])
            
          </span>
        </li>
      </ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          -5.6
        </b>
        ] (mean=[
        <b>
          -5.6
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      Defensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          3.8
        </b>
        ] (mean=[
        <b>
          3.8
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         Key lineups:
      </li>
      <ul>
        <li>
          <div>
            <a
              href="/?autoOffQuery=false&baseQuery=test&offQuery=NOT%20%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&onQuery=%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&"
              target="_blank"
            >
              AaWiggins / AnCowan / ErAyala / JaSmith
            </a>
          </div>
           contrib=[
          <b>
            3.85
          </b>
          ]  (poss=[
          <b>
            128.2
          </b>
          ], [
          <b>
            38.5
            %
          </b>
          ])
        </li>
        <li>
          Key sub-ins: 
          <span>
            <div>
              <a
                href="/TeamReport?baseQuery=%28%7B%22Wiggins%2C%20Aaron%22%3B%22DoScott%22%7D%3D1%29%20AND%20%28test%29&filter=AaWiggins&incRepOnOff=true&repOnOffDiagMode=20&"
                target="_blank"
              >
                DoScott
              </a>
            </div>
             (poss=[
            <b>
              102
            </b>
            ])
            
          </span>
        </li>
      </ul>
      <li>
        [
        <b>
          0
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          0.0
        </b>
        ] (mean=[
        <b>
          0.0
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      <a
        href="#"
        onClick={[Function]}
      >
        To clear the detailed analysis and player filter click here
      </a>
      .
    </li>
  </ul>
</span>
`;
